WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentService.swift

```diff
@@ -0,0 +1,91 @@
+
+import UIKit // TODO: remove after update to `ViewControllerPresenting` when #12864 is done
+
+final class CardPresentPaymentService: CardPresentPaymentFacade {
+    let paymentEventPublisher: AnyPublisher<CardPresentPaymentEvent, Never>
+
+    var connectedReaderPublisher: AnyPublisher<CardPresentPaymentCardReader?, Never> {
+        connectedReaderSubject.eraseToAnyPublisher()
+    }
+
+    // I don't actually think we need this, but just in case the service needs to send its own events
+    private let paymentEventSubject = PassthroughSubject<CardPresentPaymentEvent, Never>()
+    private let connectedReaderSubject = CurrentValueSubject<CardPresentPaymentCardReader?, Never>(nil)
+
+    private let onboardingAdaptor: CardPresentPaymentsOnboardingPresenterAdaptor
+
+    private let paymentAlertsPresenterAdaptor: CardPresentPaymentAlertsPresenting
+    private let connectionControllerManager: CardPresentPaymentsConnectionControllerManager
+
+    private let siteID: Int64
+
+    private var cardPresentPaymentsConfiguration: CardPresentPaymentsConfiguration {
+        CardPresentConfigurationLoader().configuration
+    }
+
+    init(siteID: Int64) {
+        self.siteID = siteID
+        let onboardingAdaptor = CardPresentPaymentsOnboardingPresenterAdaptor()
+        self.onboardingAdaptor = onboardingAdaptor
+        let paymentAlertsPresenterAdaptor = CardPresentPaymentsAlertPresenterAdaptor()
+        self.paymentAlertsPresenterAdaptor = paymentAlertsPresenterAdaptor
+        connectionControllerManager = CardPresentPaymentsConnectionControllerManager(
+            siteID: siteID,
+            configuration: CardPresentConfigurationLoader().configuration,
+            alertsPresenter: paymentAlertsPresenterAdaptor)
+
+        paymentEventPublisher = onboardingAdaptor.onboardingScreenViewModelPublisher
+            .map { onboardingEvent -> CardPresentPaymentEvent in
+                switch onboardingEvent {
+                case .showOnboarding(let onboardingViewModel):
+                    return .showOnboarding(onboardingViewModel)
+                case .onboardingComplete:
+                    return .idle
+                }
+            }
+            .merge(with: paymentAlertsPresenterAdaptor.paymentAlertPublisher)
+            .merge(with: paymentEventSubject)
+            .eraseToAnyPublisher()
+    }
+
+    @MainActor
+    func connectReader(using connectionMethod: CardReaderConnectionMethod) async throws -> CardPresentPaymentReaderConnectionResult {
+        // What happens if this gets called while there's another connection ongoing?
+        let preflightController = createPreflightController()
+
+        await preflightController.start(discoveryMethod: connectionMethod.discoveryMethod)
+
+        // TODO: replace it with reader connection
+        let mockReader = CardPresentPaymentCardReader(name: "Test Reader", batteryLevel: 0.5)
+        connectedReaderSubject.send(mockReader)
+        return .connected(mockReader)
```

Comment:
> For the next step to return the connection result, do you think we observe CardPresentPaymentPreflightControllerProtocol.readerConnection or is there another way that's more compatible with async/await?

Both a return, and the stream, are available. We should return the reader from this function (as we do with the mock here) and also publish it on the `readerConnection` stream. 

Putting it on the stream lets other things react to it, e.g. the reader status indicator in the nav bar or checkout, but the caller gets a definitive response from the return.

> For connectedReaderPublisher implementation in this payment service (right now just a placeholder subject): do you think the subject just emits the result here on every connectReader call? At first I thought about using CardReaderSettingsKnownReaderProvider, but it might be safer to recognize a reader connection from an explicit connection call using CardPresentPaymentPreflightController

Yes, but also we would want to emit a `disconnection` event if the reader is turned off using the button on the device.

The known reader provider probably isn't much use for our purposes here, it only keeps track of whether we've previously connected to a particular reader and should auto-connect to it when we discover it.

> For the UI to display the reader connection status anytime, I think connectReader would need to be called first when entering the POS mode, and then later triggered by user taps on the disconnection status.

Yeah, potentially, but only if we don't show any UI during that call. We may need a different version of the function to do it, e.g. to run preflight from init without presenting UI.

Reply:
> > For the next step to return the connection result, do you think we observe CardPresentPaymentPreflightControllerProtocol.readerConnection or is there another way that's more compatible with async/await?
> 
> Both a return, and the stream, are available. We should return the reader from this function (as we do with the mock here) and also publish it on the `readerConnection` stream.
> 
> Putting it on the stream lets other things react to it, e.g. the reader status indicator in the nav bar or checkout, but the caller gets a definitive response from the return.

Thanks for taking on this in https://github.com/woocommerce/woocommerce-ios/pull/12928!

Reply:
> Yes, but also we would want to emit a disconnection event if the reader is turned off using the button on the device.

Great point, do we have a way to observe this right now through what's available in the adaptors? `CardPresentPaymentAction.publishCardReaderConnections` looks like a place we can observe and detect the disconnection event. We can also handle this after the happy path. Also, it looks like the navigation option we're going p1717470952308339/1717083791.879079-slack-C070SJRA8DP, the card reader connection status will be in a menu that is not always visible. If this is the case, we can trigger a silent version of `connectReader` only when the menu appears.

Reply:
Yeah, that's the one.

After happy path sounds good, but I think we should probably handle it regardless of the UI path we take, to make sure we've got correctness in this class.

We'll need it on the checkout page too, IIRC, as we have a big "Reader connected" status there.

Reply:
Yup it's an important task that's worth tracking now, I added [a task](https://github.com/orgs/woocommerce/projects/247/views/6?sliceBy%5Bvalue%5D=iOS+%F0%9F%8D%8E&pane=issue&itemId=65994779) to the board.